src/lib.rs

```diff
@@ -0,0 +1,386 @@
+            })
+        } else {
+            Channel::Pro(ProChannel { handle, info })
+        })
+    }
+}
+
+pub trait UniversalCaptureFamily {
+    fn handle(&self) -> *mut c_void;
+
+    fn info(&self) -> &ChannelInfo;
+
+    fn get_video_signal_status(&self) -> Result<VideoSignalStatus> {
+        let mut status = MaybeUninit::uninit();
+        unsafe {
+            if sys::MWGetVideoSignalStatus(self.handle(), status.as_mut_ptr())
+                != sys::_MW_RESULT__MW_SUCCEEDED
+            {
+                whatever!("unable to get video signal status");
+            }
+            Ok(status.assume_init().into())
+        }
+    }
+}
+
+pub trait ProEcoCaptureFamily: UniversalCaptureFamily {
+    fn get_device_time(&self) -> Result<Duration> {
+        let mut time: c_longlong = 0;
+        unsafe {
+            if sys::MWGetDeviceTime(self.handle(), &mut time as *mut _)
+                != sys::_MW_RESULT__MW_SUCCEEDED
+            {
+                whatever!("unable to get device time");
+            }
+            Ok(Duration::from_nanos(100 * time as u64))
+        }
+    }
+}
+
+impl UniversalCaptureFamily for Channel {
+    fn handle(&self) -> *mut c_void {
+        match self {
+            Channel::Eco(ch) => ch.handle(),
+            Channel::Pro(ch) => ch.handle(),
+        }
+    }
+
+    fn info(&self) -> &ChannelInfo {
+        match self {
+            Channel::Eco(ch) => ch.info(),
+            Channel::Pro(ch) => ch.info(),
+        }
+    }
+}
+
+pub struct EcoChannel {
+    handle: *mut c_void,
+    info: ChannelInfo,
+    event_fd: EventFd,
+    // hold onto a reference of the currently set capture frame
+    video_capture_frame: Option<Pin<Box<EcoVideoCaptureFrame>>>,
+}
+
+impl UniversalCaptureFamily for EcoChannel {
+    fn handle(&self) -> *mut c_void {
+        self.handle
+    }
+
+    fn info(&self) -> &ChannelInfo {
+        &self.info
+    }
+}
+
+impl ProEcoCaptureFamily for EcoChannel {}
+
+impl EcoChannel {
+    pub fn start_video_capture(&mut self, width: u16, height: u16, format: FourCC) -> Result<()> {
+        let mut params = sys::_MWCAP_VIDEO_ECO_CAPTURE_OPEN {
+            cx: width as _,
+            cy: height as _,
+            dwFOURCC: format.as_u32(),
+            llFrameDuration: -1,
+            hEvent: self.event_fd.as_raw_fd() as _,
+        };
+        unsafe {
+            if sys::MWStartVideoEcoCapture(self.handle(), &mut params as *mut _)
+                != sys::_MW_RESULT__MW_SUCCEEDED
+            {
+                whatever!("unable to start video capture");
+            }
+        }
+        Ok(())
+    }
+
+    pub fn stop_video_capture(&mut self) -> Result<()> {
+        unsafe {
+            if sys::MWStopVideoEcoCapture(self.handle()) != sys::_MW_RESULT__MW_SUCCEEDED {
+                whatever!("unable to stop video capture");
+            }
+        }
+        Ok(())
+    }
+
+    pub fn set_video_capture_frame(&mut self, frame: EcoVideoCaptureFrame) -> Result<()> {
```

Comment:
What happens if you call this while the hardware device / some other thread created by the library is writing? I don't see a guarantee `MWCaptureSetVideoEcoFrame` will return failure before doing anything dumb, so can it continue writing after this code drops the buffer? If so, this code should at least fail if `self.video_capture_frame.is_some()`, if not bundle this with the start/wait/get operations more somehow.

Reply:
I don't know what would happen. I don't think there's a scenario where I would expect the device to write to the frame after we drop it, but I'll at least add that check.

Reply:
I mean specifically: what if you call `set_video_capture_frame` twice without a successful `wait` between?

...and another scenario occurs to me: what if you call `set_video_capture_frame` and then drop the `EcoChannel` without a successful `wait` between? I might have been wrong about not needing `impl Drop for EcoChannel` if there's an `impl Drop for ChannelHandle`...

Reply:
I'm hoping:

- If you call `MWCaptureSetVideoEcoFrame`, any frame that was previously given will no longer be used after it returns. So calling it twice in a row should be fine.
- If you drop the channel handle, all operations associated with that handle will stop before `MWCloseChannel` returns.

I don't know for sure whether these are actually the case, but I did test them and everything seemed to go as expected.